kernel/binding.cc

```diff
@@ -24,6 +24,163 @@ YOSYS_NAMESPACE_BEGIN
+	static int binding_counter;
+	attr_name = stringf("\\bound%d", binding_counter++);
+}
+
+// Return true if mod is either called name or is derived from an abstract
+// module that was called name.
+static bool
+module_original_type_matches(const RTLIL::Module &mod, RTLIL::IdString name)
+{
+	if (mod.name == name) return true;
+
+	// A module derived from 'foo' will have a name starting with $paramod
+	// and an 'hdlname' attribute equal to '\foo'.
+	if (!mod.name.begins_with("$paramod"))
+		return false;
+
+	RTLIL::IdString hdlname = mod.get_string_attribute(ID::hdlname);
+	log_assert(!hdlname.empty());
+	return (hdlname == name);
+}
+
+// Search through the design for modules called name, or that are derived from
+// one called name.
+static std::vector<RTLIL::Module *>
+find_concrete_modules(RTLIL::Design &design, RTLIL::IdString name)
+{
+	std::vector<RTLIL::Module *> mods;
+
+	for (const auto &pr : design.modules_) {
+		if (module_original_type_matches(*pr.second, name))
+			mods.push_back(pr.second);
+	}
+
+	return mods;
+}
+
+// Try to find a cell below mod whose (dot-separated) name is given by path.
+// Modify so_far in place and return true on success.
+static bool
+find_cell_in(RTLIL::Binding::Path *so_far,
+             RTLIL::Design &design, RTLIL::Module &mod, const char *path)
+{
+	const char *dot = strchr(path, '.');
+	std::string cell_name =
+		dot ? std::string(path, dot - path) : std::string(path);
+
+	RTLIL::Cell *cell = mod.cell('\\' + cell_name);
+	if (!cell)
+		return false;
+
+	so_far->push_back(RTLIL::Binding::Item(&mod, cell));
+	if (!dot)
+		return true;
+
+	// We found a child cell, but there's more searching to do. Recurse to
+	// search the cell's module.
+	return find_cell_in(so_far, design, *cell->module, dot + 1);
+}
+
+RTLIL::Binding::Path
+RTLIL::Binding::find_rel_cell(RTLIL::Design &design, RTLIL::Module &start) const
+{
+	Path ret;
+	bool found = find_cell_in(&ret, design, start, target_name.c_str() + 1);
+	if (!found)
+		return Path();
+
+	log_assert(ret.size());
+	check_cell_type(design, *ret.back().second);
+	return ret;
+}
+
+std::vector<RTLIL::Binding::Path>
+RTLIL::Binding::find_tl_cells(RTLIL::Design &design) const
+{
+	std::vector<Path> paths;
+
+	std::string tgt_name = target_name.str();
+	size_t tgt_off = 0;
+
+	// The target name should always start with a backslash
+	log_assert(tgt_name[0] == '\\');
+
+	// Is the name absolute (starts with $root)? If so, just skip past that:
+	// we're at the top already.
+	if (tgt_name.compare(0, 7, "\\$root.") == 0)
+		tgt_off = 7;
+	else
+		tgt_off = 1;
+
+	// This function can only find cells (as opposed to modules at the top
+	// level). If there's no '.' after tgt_off, give up.
+	size_t dot_off = tgt_name.find('.', tgt_off);
+	if (dot_off == std::string::npos)
+		return paths;
+
+	// Otherwise, characters with indices {tgt_off .. dot_off - 1} should
+	// name a module and then {dot_off + 1 .. } should name the cell
+	// relative to the module.
```

Comment:
For target names which contain a dot, they must be a `bind_target_instance`, as `bind_target_scope` cannot contain a dot. So in this case, the target name must "specify a single instance into which the `bind_instantiation` should be inserted."

For bind directives within modules, when the target name has a dot, other tools do not allow the leading portion of the identifier to refer to a module rather than an instance. For bind directives in the compilation unit scope, when the target name has a dot, they disallow binding into a module which is instantiated twice. I believe these interpretations of a hierarchical identifier and the `bind` spec are accurate.

So, given the following modules
```SystemVerilog
module foo;
  initial $display("hello from foo");
endmodule
module bar;
  initial $display("hello from bar");
endmodule
module act;
  initial $display("hello from act");
  foo f();
endmodule
```

I believe each of the following cases should be **illegal**:
```SytemVerilog
module beep;
  act act1();
  //act act2();
endmodule
module boop;
  beep b1();
  beep b2();
endmodule
bind act.f bar b();
```
```SytemVerilog
module beep;
  act act1();
  act act2();
endmodule
module boop;
  beep b1();
endmodule
bind act.f bar b();
```
```SytemVerilog
module beep;
  act act1();
  act act2();
  bind act.f bar b();
endmodule
```

Whereas each of the following cases are **legal**:
```SystemVerilog
module beep;
  act act1();
endmodule
module boop;
  beep b1();
endmodule
bind act.f bar b();
```
```SystemVerilog
module beep;
  act act1();
endmodule
module boop;
  beep b1();
endmodule
bind act bar b();
```
```SystemVerilog
module beep;
  act act1();
  act act2();
endmodule
module boop;
  beep b1();
endmodule
bind beep.act1.f bar b();
```
```SytemVerilog
module beep;
  act act1();
  act act2();
  bind act1.f bar b();
endmodule
```

Reply:
> For bind directives within modules, when the target name has a dot, other tools do not allow the leading portion of the identifier to refer to a module rather than an instance.

Hmm. If I understand the spec correctly, both are explicitly allowed. Page 709 of IEEE 1800-2017 (section 23.3.1) has (emphasis mine):
>> `$root` allows explicit access to the top of the instantiation tree. This is useful to disambiguate a local path
(which takes precedence) from the rooted path. If `$root` is not specified, a hierarchical path is ambiguous.
**For example, `A.B.C` can mean the local `A.B.C` or the top-level `A.B.C` (assuming there is an instance A that
contains an instance B at both the top level and in the current module).** The ambiguity is resolved by giving
priority to the local scope and thereby preventing access to the top-level path. `$root` allows explicit access
to the top level in those cases in which the name of the top-level module is insufficient to uniquely identify
the path.

In the `find_tl_cells` method (where this comment points!), we only handle the top-level case. But we only call it from `insert_binding` in `hierarchy.cc` if `find_rel_cell` failed (handling the local case). I *think* that this matches the spec?

Looking one-by-one at the cases you listed:

For (1) - (3), I think that `bind act.f bar b();` parses the same in each case. `act.f` matches the hierarchical identifier in the BNF's `bind_target_instance`. The interpretation in (1) and (2) is forced to be "treat `act` as a module and point at the `f` instance in that top-level module". In (1), we get one copy and in (2), we get 2 copies. In (3), we perform a local search for `act.f` which fails and then fall back to the top-level interpretation. (A *horrible* design, but I think it's legal...)

For (4)-(7), I agree that they are legal (and I think we probably agree about what they mean).

Reply:
You have convinced me that properly hierarchical `bind`s within a module may fall back to looking up a module rather than an instance. Hence, each of the following should and do work in this PR and in some commercial suites as well.
```SystemVerilog
module beep;
  act act1();
  bind act.f bar b();
endmodule
```
```SystemVerilog
module beep;
  act act1();
endmodule
module boop;
  beep b1();
  bind act.f bar b();
endmodule
```

> In (1), we get one copy...

There are two instances of `beep`, so in that test case, there are two instances of `act`. Indeed, there are two instances of `act` in all of (1) thru (3). I think other tools are relying on the "specify a **single instance** into which the bind_instantiation should be inserted" language to avoid having to modify multiple instances in these cases. I do think it would be a bit strange for a single `hierarchical_identifier` to referring to multiple cells.

Reply:
You're right that I can't count :-) But, ignoring that, can you explain in what way you think this code's behaviour should change? At the moment, it matches how I've understood the spec: presumably we disagree on some point there, but I'm not quite sure which one!

Reply:
The spec appears to state that a `bind_target_instance` is used to "specify a **single** instance." But in (1) thru (3) above, it is being used to specify **multiple** distinct instances. I have not been able to find any tool which accepts (1) thru (3), and I suspect the reason is because of this restriction.

Reply:
Ah, I think I understand. To implement the restriction, I guess we could propagate some "instantiation count" down from the top of the tree and only allow a hierarchical bind via module if that count happened to be 1. But it's pretty complicated and only enforces a restriction that's kind of arguable :-)

My reading of the spec's "*The second form of bind syntax can be used to specify a single instance into which the bind_instantiation should be inserted.*" is, I think, slightly different from yours. Having used `bind` in Verilator (which doesn't support this form), the annoying problem is when you have multiple instances of a module and you only want to bind into some of them. I think this form is to support that use-case. For example, suppose you wanted to do something like "bind this protocol checker into the driver for each of my AXI blocks, but not into the driver when used in in my other interfaces". There, you want a target that's something like `axi_block_module.driver_instance`.

The "just one instance in the design" interpretation would mean that's not allowed if you instantiate `axi_block_module` more than once. Instead, you'd have to write something like `$root.path.to.axi_block_instance0.driver_instance` and `$root.path.to.axi_block_instance1.driver_instance`, which is a bit horrible: it encodes the number of AXI blocks (easy to forget if you add one!) and the exact design hierarchy (causing churn when `path.to` changes).

My take is:
- The spec is slightly ambiguous here. It might really mean "just one instance in the design" or it might not.
- If it doesn't mean that, the proposed code is correct.
- If it does, the proposed code will work for any correct design and will do something reasonable if the design is invalid. Another tool might choose to choke on the invalid design, but there's no requirement for us to do so.

So I *think* I'd stick with the current code, which is in keeping with Yosys's error handling approach in general ("don't try too hard; this is a synthesis tool, not a simulator"). Of course, I wrote it, so I'm biased :-D If you disagree, do you have a suggestion for a simple way to enforce the restriction that you think should hold?

Reply:
Oh, I meant to say: you wrote: "I have not been able to find any tool which accepts (1) thru (3), and I suspect the reason is because of this restriction.". When you tried it out, what were the error messages you saw? Were they of the form "No, you fool! There are multiple instances of X" or were they "Aargh. Internal error". In the former case, that's a strong argument. In the latter, not so much... :-)

Reply:
I'm away from my computer at the moment, so I won't be able to re-run to grab those error messages. If you're interested, I often use edaplayground.com, which provides limited access to some commercial simulators for students, researchers, etc.

Regarding how to specify a limited set of modules (especially locally), rather than the following:
```SystemVerilog
module beep;
  act act1();
  act act2();
  bind act.f bar b();
endmodule
```
The spec seems to require something like:
```SystemVerilog
module beep;
  act act1();
  act act2();
  bind act1.f bar b();
  bind act2.f bar b();
endmodule
```

I think you may be right that this behavior wouldn't seem to affect definitely legal source, but I'd like to poke around a little more later today when I get the chance.

You described a rather complex strategy for implementing this restriction. I agree something that complex would be undesirable. Would it be possible to do it by erroring whenever we match multiple modules via a properly hierarchical identifier, i.e., one with dot(s)? That is the sort of restriction I had in my head, which seems like it might make certain pieces even simpler.

Reply:
Here are some relevant test results run against three different commercial simulators.

```SystemVerilog
module beep;
	act act1();
	bind act.f bar b();
endmodule
```
A) `Hierarchical name component lookup failed for 'act' at 'beep'`
B) `Failed to find 'act' in hierarchical name act.f`
C) (no complaint!)





```SystemVerilog
module beep;
	act act1();
	act act2();
	bind act.f bar b();
endmodule
```
A) `Hierarchical name component lookup failed for 'act' at 'beep'`
B) `Failed to find 'act' in hierarchical name act.f`
C) `Can only bind to modules or instances`


```SystemVerilog
module beep;
	act act1();
endmodule
bind act.f bar b();
```
A) `Hierarchical name component lookup failed for 'act' at '$unit_0x4ccdf83b'`
B) `Failed to find 'act' in hierarchical name act.f`
C) no error



```SystemVerilog
module beep;
	act act1();
	act act2();
endmodule
bind act.f bar b();
```
A) `Hierarchical name component lookup failed for 'act' at '$unit_0x4ccdf83b'`
B) `Failed to find 'act' in hierarchical name act.f`
C) `Can only bind to modules or instances`

Reply:
Ahah! I think I've finally understood what's going on. It turns out that the important bit is the previous paragraph of section 23.3.1 (at the bottom of p708). The point is that `my_module.a.b.c` is only allowed if `my_module` is a "top-level module".

A top-level module is a module that isn't instantiated by any other module. In particular, this will guarantee that there's at most one instance of the thing that the hierarchical reference points to.

I'll fix up the code accordingly, and include your examples from above in the tests.